crates/isograph_compiler/src/source_files.rs

```diff
@@ -0,0 +1,370 @@
+use graphql_lang_types::{GraphQLTypeSystemDocument, GraphQLTypeSystemExtensionDocument};
+use graphql_schema_parser::{parse_schema, parse_schema_extensions};
+use intern::string_key::Intern;
+use isograph_config::CompilerConfig;
+use isograph_lang_parser::IsoLiteralExtractionResult;
+use isograph_schema::UnvalidatedSchema;
+
+use crate::{
+    batch_compile::BatchCompileError,
+    isograph_literals::{
+        process_iso_literals, read_and_parse_iso_literals, read_file, read_files_in_folder,
+    },
+    refetch_fields::add_refetch_fields_to_objects,
+    schema::read_schema_file,
+    watch::{ChangedFileKind, SourceEventKind, SourceFileEvent},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq, Default)]
+pub struct SourceFiles {
+    pub schema: GraphQLTypeSystemDocument,
+    pub schema_extensions: HashMap<SourceFileName, GraphQLTypeSystemExtensionDocument>,
+    pub contains_iso: ContainsIso,
+}
+
+impl SourceFiles {
+    pub fn new(config: &CompilerConfig) -> Result<Self, BatchCompileError> {
+        let schema = read_and_parse_graphql_schema(&config.schema)?;
+
+        let mut schema_extensions = HashMap::new();
+        for schema_extension_path in config.schema_extensions.iter() {
+            let (file_path, extensions_document) =
+                read_and_parse_schema_extensions(schema_extension_path)?;
+            schema_extensions.insert(file_path, extensions_document);
+        }
+
+        let mut contains_iso = ContainsIso::default();
+        read_and_parse_iso_literals_from_folder(
+            &mut contains_iso,
+            &config.project_root,
+            &config.project_root,
+        )?;
+
+        Ok(Self {
+            schema,
+            schema_extensions,
+            contains_iso,
+        })
+    }
+
+    pub fn create_unvalidated_schema(
+        self,
+        schema: &mut UnvalidatedSchema,
+        config: &CompilerConfig,
+    ) -> Result<(), BatchCompileError> {
+        let outcome = schema.process_graphql_type_system_document(self.schema, config.options)?;
+        for extension_document in self.schema_extensions.into_values() {
+            let _extension_outcome = schema
+                .process_graphql_type_extension_document(extension_document, config.options)?;
+        }
+        process_iso_literals(schema, self.contains_iso)?;
+        process_exposed_fields(schema)?;
+        schema.add_fields_to_subtypes(&outcome.type_refinement_maps.supertype_to_subtype_map)?;
+        schema
+            .add_pointers_to_supertypes(&outcome.type_refinement_maps.subtype_to_supertype_map)?;
+        add_refetch_fields_to_objects(schema)?;
+        Ok(())
+    }
+
+    pub fn update(
+        &mut self,
+        config: &CompilerConfig,
+        changes: &[SourceFileEvent],
+    ) -> Result<(), BatchCompileError> {
+        let errors = changes
+            .iter()
+            .filter_map(|(event, change_kind)| match change_kind {
+                ChangedFileKind::Config => None,
```

Comment:
should we panic here? Encountering a config change is unexpected.

Reply:
yeah, for now it means a bug. I thought eventually we can handle some of the changes without a full compilation but we should panic atm